Documentation - add integration guide (#16)
commercetools/commercetools-subscriptions: docs/IntegrationGuide.md

  - [Subscription order custom fields](#subscription-order-custom-fields)
- [Non-recoverable error](#non-recoverable-error)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Terminology

- **user**: the one who is responsible for running `commercetools-subscriptions` and `frontend`.
- **customer**: the one who buy goods from the user.
- **checkout-order**: an order which customers create during the checkout. It may contain one or more subscriptions. `commercetools-subscriptions` will generate a **template-order** for every subscription.
- **frontend**: application (usually an e-shop) of a user that is responsible for generating checkout orders.
- **template-order**: an order which manages a single subscription and is used as a template to create a
  subscription-order.
- **subscription-order**: a new order (actual delivery) triggered by the template order according to the defined schedule.

## Before you begin

Create all the required custom types and states located in [resources folder](./resources). Check [project requirements](./docs/HowToRun.md#commercetools-project-requirements) for more details.

## Sequence diagram

> This diagram contains all the interactions between participants frontend, commercetools subscriptions and commercetools platform. For frontend integration only the interactions noted with Step 1, Step 2 and Step 3 are required.

Reviewer: ```suggestion
This diagram contains all the interactions between participants' frontend, commercetools subscriptions and commercetools platform. For frontend integration only the interactions noted with Step 1, Step 2 and Step 3 are required.
```

```mermaid
sequenceDiagram
    participant frontend
    participant commercetools platform
    participant commercetools subscriptions
    rect rgb(191, 223, 255)
    frontend->>+commercetools platform: Create checkout orders (Step 1)
    commercetools subscriptions->>+commercetools platform: Get checkout orders
    commercetools platform->>-commercetools subscriptions: Checkout orders
    commercetools subscriptions->>commercetools platform: Create template orders
    end
    rect rgb(204, 255, 255)
    commercetools subscriptions-->>+commercetools platform: Set template order state to `SendReminder`
    frontend-->>+commercetools platform: Get template orders with state `SendReminder` (Step 2)

>     commercetools platform-->>-frontend: Template orders to send reminder (Step 2)

Reviewer: ```suggestion
    commercetools platform-->>-frontend: Template orders to send reminders (Step 2)
```

> - If the checkout-order was placed until 15 of January then the next subscription-order creation is planned for the 1st of February.

Reviewer: cuttoffDays -> If the `amount` of days since placing `of` the checkout-order -> If the `number` of days since placing the checkout-order...

- [Create a Subscription](https://docs.commercetools.com/api/projects/subscriptions#create-a-subscription) for messages of type [`OrderStateTransition`](https://docs.commercetools.com/api/message-types#orderstatetransitionmessage)
- Query for the messages of type `OrderStateTransition` using [Messages endpoint](https://docs.commercetools.com/api/projects/messages)

Be aware that `commercetools-subscriptions` do not send any reminders, it only set custom state of the order.

After user processes the message, user has to [make a transition](https://docs.commercetools.com/api/projects/orders#transition-state) to the new order custom state `key="commercetools-subscriptions-reminderSent"`

## Step 3: Generate a subscription order

`commercetools-subscriptions` sets `nextDeliveryDate` when generating the template order. When `nextDeliveryDate` is the current date, `commercetools-subscriptions` will make a **POST** request to the URL sets using env var `SUBSCRIPTION_ORDER_CREATION_URL` with the following body payload:

```json
{ "templateOrderId": "id-of-the-template-order" }
```

> The receiver of the body payload uses `templateOrderId` to fetch the template order and create a subscription order. The subscription order must have generated unique `orderNumber`, correct payments and setting all the information required by the merchant for the further processing. For subscription order creation we recommend using [Order import API](https://docs.commercetools.com/api/projects/orders-import#orderimportdraft). On order creation make sure you have a successful payment and required fields `subscriptionTemplateOrderRef` and `deliveryDate` (value copied from the nextDeliveryDate) from the template order are set.

Reviewer: ```suggestion
The receiver of the body payload uses `templateOrderId` to fetch the template order and create a subscription order. The subscription order must have generated a unique `orderNumber`, correct payments and set all the information required by the merchant for the further processing. For subscription order creation we recommend using [Order import API](https://docs.commercetools.com/api/projects/orders-import#orderimportdraft). On order creation make sure you have a successful payment and the required fields `subscriptionTemplateOrderRef` and `deliveryDate` (value copied from the nextDeliveryDate) from the template order are set.
```

> After finishing the subscription order creation process, the receiver of the payload must return one of the following HTTP code depending on the result of the subscription order creation process:

Reviewer: ```suggestion
After finishing the subscription order creation process, the receiver of the payload must return one of the following HTTP status codes depending on the result of the subscription order creation process:
```

| Name                         | Type      | Description                                                                                                                                                      | Required |
| ---------------------------- | --------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------- |
| subscriptionTemplateOrderRef | Reference | Reference to the template order                                                                                                                                  | YES      |
| deliveryDate                 | Date      | Copy of the nextDeliveryDate from the template order. The purpose of this field is to avoid (in some error cases) the creation of duplicate subscription orders. | YES      |

## Non-recoverable error

Following error message indicates a situation when recovery cannot be done by retrying.

> Template order ${orderNumber}: Unrecoverable error received when calling ${subscriptionOrderCreationUrl}. Please check the order and set its state back to "Active". Response status code ${response.status}

> This error can happen when calling the URL sets using the env var `SUBSCRIPTION_ORDER_CREATION_URL`. According to the returned HTTP Status code following issues could happen:

Reviewer: ```suggestion
This error can happen when calling the URL sets using the env var `SUBSCRIPTION_ORDER_CREATION_URL`. According to the returned HTTP Status code the following issues might happen:
```

Author: It's the same meaning here.

| HTTP Status code | Possible error                                             | Possible solution                                                                                                                                                                    |
| ---------------- | ---------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |

> | 5xx              | Error inside the application generating subscription order | Check for errors in your application that is serving response for `SUBSCRIPTION_ORDER_CREATION_URL`                                                                                  |

Reviewer: ```suggestion
| 5xx              | Error inside the application generating subscription order | Check for errors in your application that is serving a response for `SUBSCRIPTION_ORDER_CREATION_URL`                                                                                  |
```
